frontend/lib/parsing/parsing-queries.cpp

```diff
@@ -100,25 +102,226 @@ static Parser helpMakeParser(Context* context,
+  ser.write((uint64_t)paths.size());
+
+  std::vector<std::pair<std::string, std::string>> data;
+  uint64_t offset = 0;
+
+  // Use the same serializer so that we can build up a unified cache of
+  // UniqueStrings for this library file.
+  //
+  // Store all the text in a stringstream to be written out after the cache
+  // is written.
+  std::stringstream ss;
+  chpl::Serializer builderSer(ss);
+  for (auto path : paths) {
+    UniqueString empty;
+    auto& result = parseFileToBuilderResult(context, path, empty);
+    ss.str(std::string()); // clear for this iteration
+    result.serialize(builderSer);
+
+    const auto& str = ss.str();
+    data.push_back({path.str(), str});
+
+    // write the filename and the offset for the header
+    ser.write(path.str());
+    ser.write(offset);
+    offset += str.size();
+  }
+
+  const auto& stringCache = builderSer.stringCache();
+
+  ser.write((uint32_t)stringCache.size());
+  for (const auto& kv : stringCache) {
+    const auto& pair = kv.second;
+    ser.write(pair.first); // unique ID in this table
+    ser.write((uint32_t)pair.second); // string size
+    if (pair.second > 0) {
+      // string data
+      ser.os().write(kv.first, pair.second);
+    }
+  }
+
+  // Finally, write the saved strings from serializing a BuilderResult
+  for (const auto& pair : data) {
+    ser.os().write(pair.second.c_str(), pair.second.size());
+  }
+}
+
+LibraryFile::LibraryFile(Context* context, UniqueString libPath)
+: path_(libPath) {
+  std::ifstream myFile;
+  myFile.open(libPath.str(), std::ios::in | std::ios::binary);
+  chpl::Deserializer des(context, myFile);
+
+  // Some basic validation
+  CHPL_ASSERT(LIBRARY_MAGIC == des.read<uint64_t>());
+  CHPL_ASSERT(LIBRARY_VERSION_MAJOR == des.read<int>());
+  CHPL_ASSERT(LIBRARY_VERSION_MINOR == des.read<int>());
+
+  // Currently no checking is done for 'chpl' version
+  std::ignore = des.read<int>(); // major version
+  std::ignore = des.read<int>(); // minor version
+  std::ignore = des.read<int>(); // update version
+
+  // Is this a user module, or a standard module?
+  const auto kind = des.read<std::string>();
+  CHPL_ASSERT(kind == "USER" || kind == "STANDARD");
+  isUser_ = (kind == "USER");
+
+  // Number of builder result entries
+  const auto num = des.read<uint64_t>();
+
+  // Read in the table of '.chpl' filenames and their offsets in the file
+  std::vector<std::pair<std::string, uint64_t>> offsetsTable;
+  for (uint64_t i = 0; i < num; i++) {
+    auto path = des.read<std::string>();
+    auto offset = des.read<uint64_t>();
+    offsetsTable.push_back({path, offset});
+  }
+
+  // Read unique strings
+  {
+    const uint64_t size = des.read<uint32_t>();
+    cache_.resize(size);
+    for (uint64_t i = 0; i < size; i++) {
+      int id = des.read<int>();
+      auto len = des.read<uint32_t>();
+      if (len > 0) {
+        // TODO: Can we save some memory allocations by doing this read while
+        // we're trying to create the unique c-string?
```

Comment:
👍 Another option could be to walk the entries and compute the max string size to allocate just one scratch buffer (or compute and store the max string size in the file as a`u64` before the start of the entries), or if we know the strings are all null terminated and the deserializer has a memory view, could just point into the view. Or a third potentially silly idea is to just sort the cache by length so that the longest string is first, and that provides the single buffer.